designs/cloudstack-multiple-endpoints.md

```diff
@@ -0,0 +1,177 @@
+
+```
+[Global]
+api-key    = redacted
+secret-key = redacted
+api-url    = http://172.16.0.1:8080/client/api
+```
+
+We would propose an extension of the above input mechanism so the user could provide credentials across multiple Cloudstack API endpoints like
+
+```
+[FailureDomain1]
+api-key    = redacted
+secret-key = redacted
+api-url    = http://172.16.0.1:8080/client/api
+
+[FailureDomain2]
+api-key    = redacted
+secret-key = redacted
+api-url    = http://172.16.0.2:8080/client/api
+
+[FailureDomain3]
+api-key    = redacted
+secret-key = redacted
+api-url    = http://172.16.0.3:8080/client/api
+
+...
+```
+
+We are also exploring converting the ini file to a yaml input file which contains a list of credentials and their associated endpoints. Either way, this environment variable would
+be passed along to CAPC and used by the CAPC controller just like it is currently.
+
+### Backwards Compatibility
+
+Our customer currently has clusters running with the old resource definition. In order to support backwards compatibility in the CloudstackDatacenterConfig resource, we can
+1. Make all the fields optional and see if customers have the old fields set or the new ones
+2. Introduce an eks-a version bump with conversion webhooks
+
+Between these two approaches, I would take the first and then deprecate the legacy fields in a subsequent release to simplify the code paths.
+
+However, given that the Cloudstack credentials are persisted in a write-once secret on the cluster, upgrading existing clusters may not be feasible unless CAPC supports overwriting that secret.
+
+## User Experience
+
+
+## Security
+
+The main change regarding security is the additional credential management. Otherwise, we are doing exactly the same operations - preflight check with cloudmonkey,
+create yaml templates and apply them, and then read/write eks-a resources in the eks-a controller. The corresponding change is an extension of an existing mechanism
+and there should not be any new surface area for risks than there was previously.
+
+## Testing
+
+The new code will be covered by unit and e2e tests, and the e2e framework will be extended to support cluster creation across multiple Cloudstack API endpoints.
+
+The following e2e test will be added:
+
+simple flow cluster creation/deletion across multiple Cloudstack API endpoints:
+
+* create a management+workload cluster spanning multiple Cloudstack API endpoints
+* delete cluster
+
+## Other approaches explored
+
+Another direction we can go to support this feature is to refactor the entire EKS-A codebase so that instead of all the failure domains existing inside the CloudstackDatacenterConfig
+object, each CloudstackDatacenterConfig itself corresponds with a single failure domain. Then, the top level EKS-A Cluster object could be refactored to have a list of DatacenterRefs instead
```

Comment:
It may actually provide value as we talked about wanting to support multiple environments in the future. However, I can see why it's a bigger step with less data for a specific provider.